internal/core/window.rs

```diff
@@ -1069,40 +1077,72 @@ impl WindowInner {
-    pub fn close_popup(&self) {
-        if let Some(current_popup) = self.active_popups.borrow_mut().pop() {
-            match current_popup.location {
-                PopupWindowLocation::ChildWindow(offset) => {
-                    // Refresh the area that was previously covered by the popup.
-                    let popup_region = crate::properties::evaluate_no_tracking(|| {
-                        let popup_component = ItemTreeRc::borrow_pin(&current_popup.component);
-                        popup_component.as_ref().item_geometry(0)
-                    })
-                    .translate(offset.to_vector());
-
-                    if !popup_region.is_empty() {
-                        let window_adapter = self.window_adapter();
-                        window_adapter.renderer().mark_dirty_region(popup_region.into());
-                        window_adapter.request_redraw();
-                    }
-                }
-                PopupWindowLocation::TopLevel(adapter) => {
-                    let _ = adapter.set_visible(false);
+    // Remove the popup at the given index from the popups vector.
+    fn close_popup_by_index(&self, popup_index: usize) {
+        let mut active_popups = self.active_popups.borrow_mut();
+        let current_popup = active_popups.get(popup_index).unwrap();
+
+        match &current_popup.location {
+            PopupWindowLocation::ChildWindow(offset) => {
+                // Refresh the area that was previously covered by the popup.
+                let popup_region = crate::properties::evaluate_no_tracking(|| {
+                    let popup_component = ItemTreeRc::borrow_pin(&current_popup.component);
+                    popup_component.as_ref().item_geometry(0)
+                })
+                .translate(offset.to_vector());
+
+                if !popup_region.is_empty() {
+                    let window_adapter = self.window_adapter();
+                    window_adapter.renderer().mark_dirty_region(popup_region.into());
+                    window_adapter.request_redraw();
                 }
             }
+            PopupWindowLocation::TopLevel(adapter) => {
+                let _ = adapter.set_visible(false);
+            }
         }
+
+        active_popups.remove(popup_index);
     }
 
-    /// Returns the close policy of the active popup. PopupClosePolicy::NoAutoClose if there is no active popup.
-    pub fn close_policy(&self) -> PopupClosePolicy {
+    /// Removes the popup matching the given ID.
+    pub fn close_popup(&self, popup_id: u32) {
+        let maybe_index =
+            self.active_popups.borrow().iter().position(|popup| popup.popup_id == popup_id);
+
+        if let Some(popup_index) = maybe_index {
+            self.close_popup_by_index(popup_index);
+        }
+    }
+
+    /// Close all active popups.
+    pub fn close_all_popups(&self) {
+        for (i, _) in self.active_popups.borrow().iter().enumerate() {
+            self.close_popup_by_index(i);
```

Comment:
Is that working? I fear that active_popups is still borrowed when this function is called and the borrow_mut would fail.

I can suggest having such a function `fn close_popup_impl(&slef, popup: PopupWindow)`
and then just do  `for popup in self.active_popups.take() { self.close_popup_impl(popup)   } `  in this function,   and `if let Some(popup) = self.active_popup.borrow_mut().pop() self.close_popup_impl(popup }` from close_top_popup().  I guess you get the idea.

Reply:
Thanks for catching that! It should be fixed now.